pandas/core/generic.py

```diff
@@ -561,7 +561,8 @@ def set_axis(self, labels, axis=0, inplace=False):
             obj.set_axis(labels, axis=axis, inplace=True)
             return obj
 
-    def _set_axis(self, axis, labels) -> None:
+    def _set_axis(self, axis: int, labels: Index) -> None:
```

Comment:
labels should be the same as index_like parameter of ensure_index

Reply:
thoughts on Best Practices here?  I'd like to push up the stack where we do ensure_index.  so in this case the annotation is kind of "you should pass Index, but we're not that strict about it"

Reply:
it makes sense to limit the types of internal functions to reduce the number of duplicated checks. maybe add a TODO comment here that the ensure_index added is to be eventually removed

Reply:
> "you should pass Index, but we're not that strict about it"

mypy will be

Reply:
> mypy will be

im skeptical in this case because it goes through cython.  This gets called when you do `frame.columns = ["foo"]`

Reply:
> im skeptical in this case because it goes through cython

fair enough, cython is another pain point for us with type annotations.